### PR TITLE
Persistent Queue Fix

### PR DESCRIPTION
**Pull request by airwareic**
Fixes:
- Added configuration setting to disable deletion of queue items
- Changed deletion settings of queue items for non-persistentQueue settings to 7 days and 10 attempts

Bug Fixes:
- Changed dependency resolution for IsolatedStorage to get around registration exception thrown

- Various readability and simplification improvements, especially around upload attempt counting and file locking in the queue

### CODE REVIEW

File: Source/Extras/Extensions/ExceptionlessExtraConfigurationExtensions.cs [hunk 1]
```diff
@@ -29,7 +29,7 @@ public static class ExceptionlessExtraConfigurationExtensions {
         }
 
         public static void UseIsolatedStorage(this ExceptionlessConfiguration config) {
-            config.Resolver.Register<IObjectStorage, IsolatedStorageObjectStorage>();
+            config.Resolver.Register<IObjectStorage>(new IsolatedStorageObjectStorage(config.Resolver));
```
Comment (niemyjski) [inline Source/Extras/Extensions/ExceptionlessExtraConfigurationExtensions.cs]: I don't like this change as it's been working. I'd like to have the stack trace as well as steps to reproduce this core issue before changing this.

File: Source/Extras/Exceptionless.Extras.csproj [hunk 2]
```diff
@@ -42,6 +42,7 @@
     <Reference Include="System" />
     <Reference Include="System.Configuration" />
     <Reference Include="System.Core" />
+    <Reference Include="System.Net" />
```
Comment (niemyjski) [inline Source/Extras/Exceptionless.Extras.csproj]: Although this is more readable, all changes related to adding a reference to System.Net for readability need to be reverted. We initially went down this path but it doesn't work with pcl and some pcl configurations. There are two different versions of System.NET and there are runtime errors that can occur by using types against two different runtime versions. This is why we used int status codes.

File: Source/Shared/Queue/DefaultEventQueue.cs [hunk 3]
```diff
@@ -74,35 +78,51 @@ public class DefaultEventQueue : IEventQueue {
...
-                        } else if (response.RequestEntityTooLarge) {
-                            if (batchSize > 1) {
-                                _log.Error(typeof(DefaultEventQueue), "Event submission discarded for being too large. The event will be retried with a smaller batch size.");
-                                batchSize = Math.Max(1, (int)Math.Round(batchSize / 1.5d, 0));
-                                deleteBatch = false;
-                            } else {
-                                _log.Error(typeof(DefaultEventQueue), "Event submission discarded for being too large. The event will not be submitted.");
+                        }
+                        else {
+                            switch ((HttpStatusCode)response.StatusCode)
```
Comment (niemyjski) [inline Source/Shared/Queue/DefaultEventQueue.cs]: I don't see any differences in this code other than the status code enum which we can't use due to the previous reasons :(. Can you revert this?

File: Source/Shared/Submission/SubmissionResponse.cs [hunk 4]
```diff
@@ -3,11 +3,11 @@
 
 namespace Exceptionless.Submission {
     public class SubmissionResponse {
-        public SubmissionResponse(int statusCode, string message = null) {
+        public SubmissionResponse(int statusCode, string message) {
```
Comment (niemyjski) [inline Source/Shared/Submission/SubmissionResponse.cs]: Any reason for not making this nullable?

File: Source/Shared/Extensions/FileStorageExtensions.cs [hunk 5]
```diff
@@ -3,135 +3,310 @@
...
-        public static void DeleteBatch(this IObjectStorage storage, IList<Tuple<ObjectInfo, Event>> batch) {
+        public static void DeleteBatch(this IObjectStorage storage, IList<Tuple<ObjectInfo, Event>> batch)
+        {
             foreach (var item in batch)
                 storage.DeleteObject(item.Item1.Path);
         }
 
-        public static void ReleaseBatch(this IObjectStorage storage, IList<Tuple<ObjectInfo, Event>> batch) {
+        public static void ReleaseBatch(this IObjectStorage storage, IList<Tuple<ObjectInfo, Event>> batch)
+        {
```
Comment (niemyjski) [inline Source/Shared/Extensions/FileStorageExtensions.cs]: Can you please revert all changes that are not related to the issue. I'm not trying to be nit picky and am greatful that users are doing pull requests but all our code shares style patterns and bracket changes makes things inconsistent and harder to look at the overall changes.

File: Source/Shared/Extensions/FileStorageExtensions.cs [hunk 6]
```diff
@@ -3,135 +3,310 @@
...
         }
 
-        public static void ReleaseStaleLocks(this IObjectStorage storage, string queueName, TimeSpan? maxLockAge = null) {
+        public static void ReleaseStaleLocks(this IObjectStorage storage, string queueName, TimeSpan? maxLockAge = null)
+        {
             if (!maxLockAge.HasValue)
                 maxLockAge = TimeSpan.FromMinutes(60);
 
-            foreach (var file in storage.GetObjectList(queueName + "\\q\\*.x", 500).ToList().Where(f => f.Modified < DateTime.Now.Subtract(maxLockAge.Value)))
+            foreach (var file in storage.GetObjectList(GetAllLockedQueueItemsFilter(queueName)).Where(f => f.Modified < DateTime.Now.Subtract(maxLockAge.Value)))
```
Comment (niemyjski) [inline Source/Shared/Extensions/FileStorageExtensions.cs]: We were getting at max 500 items here.

File: Source/Extras/Extensions/ExceptionlessExtraConfigurationExtensions.cs (hunk 1 above)
Comment (ejsmith) [inline Source/Extras/Extensions/ExceptionlessExtraConfigurationExtensions.cs]: Did you see an issue with this?

File: Source/Shared/Extensions/FileStorageExtensions.cs (hunk 5 above)
Comment (ejsmith) [inline Source/Shared/Extensions/FileStorageExtensions.cs]: Yes, we need the code style to match the rest of the project. It also makes it hard to see what has changed when there is a lot code style changes in the commit.

**Comment by ejsmith** [comment]
Thank you for the contribution! I really appreciate you taking the time to do this. So the primary thing that you are trying to change is to make sure that we are not throwing away any events? Have you seen events being thrown away?  Let's discuss this more and thing figure out a plan of attack that we agree on.

File: Source/Extras/Extensions/ExceptionlessExtraConfigurationExtensions.cs (hunk 1 above)
Comment (niemyjski) [inline Source/Extras/Extensions/ExceptionlessExtraConfigurationExtensions.cs]: @airwareic was seeing a TinyIOC exception but I never saw the stack trace or have seen other reports. I'd love to see the stack trace to figure out what's going on.
